transport/src/main/java/io/netty/channel/socket/oio/OioSocketChannel.java

```diff
@@ -143,24 +156,82 @@ protected int doReadBytes(ByteBuf buf) throws Exception {
     }
 
     @Override
-    public ChannelFuture shutdownOutput(final ChannelPromise future) {
+    public ChannelFuture shutdownOutput(final ChannelPromise promise) {
+        EventLoop loop = eventLoop();
+        if (loop.inEventLoop()) {
+            try {
+                socket.shutdownOutput();
+                promise.setSuccess();
+            } catch (Throwable t) {
+                promise.setFailure(t);
+            }
```

Comment:
Isn't `inEventLoop()` relatively an expensive operation? I'd prefer not letting it called twice when we can avoid.